shared/domains/chelonia/chelonia.js

```diff
@@ -1642,50 +1646,116 @@ export default (sbp('sbp/selectors/register', {
-        meta: key
-      })
-      const response = await this.config.fetch(`${this.config.connectionURL}/kv/${encodeURIComponent(contractID)}/${encodeURIComponent(key)}`, {
-        headers: new Headers([[
-          'authorization', buildShelterAuthorizationHeader.call(this, contractID)
-        ]]),
-        method: 'POST',
-        body: JSON.stringify(serializedData),
-        signal: this.abortController.signal
-      })
+      let response: Response
+      if (data || typeof onconflict !== 'function') {
+        const serializedData = outputEncryptedOrUnencryptedMessage.call(this, {
+          contractID,
+          innerSigningKeyId,
+          encryptionKeyId,
+          signingKeyId,
+          data,
+          meta: key
+        })
+        response = await this.config.fetch(url, {
+          headers: new Headers([[
+            'authorization', buildShelterAuthorizationHeader.call(this, contractID)
+          ], [
+            'if-match', ifMatch || '""'
+          ]
+          ]),
+          method: 'POST',
+          body: JSON.stringify(serializedData),
+          signal: this.abortController.signal
+        })
+      } else {
+        // If no initial data provided, perform a GET `fetch` to get the current
+        // data and CID
+        response = await this.config.fetch(url, {
+          headers: new Headers([[
+            'authorization', buildShelterAuthorizationHeader.call(this, contractID)
+          ]]),
+          signal: this.abortController.signal
+        })
+      }
+      // The `resolveData` function is tasked with computing merged data, as in
+      // merging the existing stored values (after a conflict or initial fetch)
+      // and new data. The return value indicates whether there should be a new
+      // attempt at storing updated data (if `true`) or not (if `false`)
+      const resolveData = async () => {
+        let currentValue
+        if (response.ok || response.status === 409 || response.status === 412) {
+          const serializedData = await response.json()
+          currentValue = parseEncryptedOrUnencryptedMessage.call(this, {
+            contractID,
+            serializedData,
+            meta: key
+          })
+        } else if (response.status !== 404 && response.status !== 410) {
+          throw new ChelErrorUnexpectedHttpResponseCode('[kv/set] Invalid response code: ' + response.status)
+        }
+        const result = await onconflict({
+          contractID,
+          key,
+          failedData: data,
+          status: response.status,
+          etag: response.headers.get('x-cid') || response.headers.get('etag'),
+          get currentData () {
+            return currentValue?.data
+          },
+          currentValue
+        })
+        if (!result) return false
+
+        data = result[0]
+        ifMatch = result[1]
+        return true
+      }
+      if (!data && typeof onconflict === 'function') {
+        if (await resolveData()) {
+          continue
```

Comment:
This code is extremely confusing to me.

Above we check `if (!data` and we enter into here. In most cases `resolveData()` will result in `data` being set. Then we loop again as a result of the `continue`, but from that point on this code will never be run again because `!data` will be `false`.... Is that right? If so, why is it being skipped from that point on? (A comment is needed). If not, a bug.

Reply:
That'd be for the initial iteration. Later on, there's a similar code in the `if` that follows